### PR TITLE
Typescript upgrade

### PR DESCRIPTION
**Pull request by Lumyo**
Upgrade the repository to use typescript + typechain for fully typed tests

### CODE REVIEW

File: contracts/signature/ERC1271Stored.sol [hunk 1]
```diff
@@ -11,7 +11,7 @@ import './ERC1271StoredStorage.sol';
 abstract contract ERC1271Stored is ERC1271Base {
   function _isValidSignature (
     bytes32 hash,
-    bytes memory
+    bytes memory signature
```
Comment (ItsNickBarry) [inline contracts/signature/ERC1271Stored.sol:14]: Why name the variable?  It's unused and therefore will cause a compilation warning.

File: contracts/signature/ERC1271StoredMock.sol [hunk 2]
```diff
@@ -8,7 +8,7 @@ import './ERC1271Stored.sol';
 contract ERC1271StoredMock is ERC1271Stored {
   constructor (
     bytes32 hash,
-    bytes memory
+    bytes memory signature
```
Comment (ItsNickBarry) [inline contracts/signature/ERC1271StoredMock.sol:11]: Why name the variable? It's unused and therefore will cause a compilation warning.

File: copyAbi.ts [hunk 3]
```diff
@@ -0,0 +1,237 @@
+import fs from 'fs';
```
Comment (ItsNickBarry) [inline copyAbi.ts:1]: What is the purpose of this?  Can it be accomplished using [hardhat-abi-exporter](https://github.com/ItsNickBarry/hardhat-abi-exporter)?

Comment (ItsNickBarry) [reply copyAbi.ts:1]: If not, let's move this to a post-compilation hook.  I don't want to have to modify this list every time a contract is added or removed.

File: hardhat.config.ts [hunk 4]
```diff
@@ -43,9 +44,9 @@ module.exports = {
   networks: {
     generic: {
       // set URL for external network
-      url: `${ process.env.URL }`,
+      url: `${process.env.URL}`,
```
Comment (ItsNickBarry) [inline hardhat.config.ts:47]: Going to need padding inside of these brackets.  Let's set a prettier or eslint rule at some point.

File: package.json [hunk 5]
```diff
@@ -3,19 +3,33 @@
   "private": true,
   "author": "Nick Barry",
   "license": "MIT",
+  "scripts": {
+    "compile": "hardhat compile && yarn copyAbi && yarn typechain",
```
Comment (ItsNickBarry) [inline package.json:7]: Prefer to override the compile task so that the native `hh compile` functions as expected.

File: package.json [hunk 6]
```diff
@@ -3,19 +3,33 @@
...
     "@nomiclabs/hardhat-ethers": "^2.0.2",
     "@nomiclabs/hardhat-waffle": "^2.0.1",
     "@solidstate/spec": "link:spec",
+    "@typechain/ethers-v5": "^7.0.0",
+    "@types/chai": "^4.2.18",
+    "@types/mocha": "^8.2.2",
+    "@types/node": "^15.6.0",
+    "@types/rimraf": "^3.0.0",
     "chai": "^4.3.4",
-    "eslint": "^7.24.0",
```
Comment (ItsNickBarry) [inline package.json:11]: Why remove eslint?

File: package.json [hunk 7]
```diff
@@ -3,19 +3,33 @@
...
-    "eslint-plugin-mocha": "^8.1.0",
```
Comment (ItsNickBarry) [inline package.json:12]: If we're removing this, need to remove many instances of `// eslint-disable-next-line`.

File: copyAbi.ts (hunk 3 above)
Comment (ItsNickBarry) [reply copyAbi.ts:1]: Note that ABIs are already exported to `abi` directory and published as `@solidstate/abi`.

File: test/factory/CloneFactory.ts [hunk 8]
```diff
@@ -23,10 +25,8 @@ describe('CloneFactory', function () {
 
           await instance['deployClone()']();
 
-          expect(
-            await ethers.provider.getCode(address)
-          ).to.equal(
-            await ethers.provider.getCode(instance.address)
+          expect(await ethers.provider.getCode(address)).to.equal(
```
Comment (ItsNickBarry) [inline test/factory/CloneFactory.ts:28]: This format is painful to read.  The work of `prettier`, I assume?

File: spec/token/ERC20/ERC20Base.behavior.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    let receiver: SignerWithAddress;
+    let sender: SignerWithAddress;
+    let instance: ERC20Base;
+
+    before(async function () {
+      [holder, spender, receiver, sender] = await ethers.getSigners();
+    });
+
+    beforeEach(async function () {
+      instance = await deploy();
```
Comment (ItsNickBarry) [inline spec/token/ERC20/ERC20Base.behavior.ts:33]: How does Typescript handle casting of contracts?  If a third party contract has both `balanceOf(address)` and a `balanceOf(address,uint256)` functions, for example, will the calls in this file fail due to the overload?

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+    "compilerOptions": {
+        "target": "es2018",
+        "module": "commonjs",
+        "strict": true,
+        "esModuleInterop": true,
+        "outDir": "dist"
+    },
+    "include": ["./scripts", "./test"],
+    "files": ["./hardhat.config.ts"]
+}
```
Comment (ItsNickBarry) [inline tsconfig.json:11]: Many files lack newlines, and Github complains.

File: contracts/signature/ERC1271Stored.sol (hunk 1 above)
Comment (Lumyo) [reply contracts/signature/ERC1271Stored.sol:14]: Not having this variable named prevents typechain from generating the typed contracts.

image

File: contracts/signature/ERC1271StoredMock.sol (hunk 2 above)
Comment (Lumyo) [reply contracts/signature/ERC1271StoredMock.sol:11]: Not having this variable named prevents typechain from generating the typed contracts.

image

File: copyAbi.ts (hunk 3 above)
Comment (Lumyo) [reply copyAbi.ts:1]: It requires the full artifacts.
I just found out about this hardhat plugin : https://www.npmjs.com/package/hardhat-typechain
I'll replace with that

File: package.json (hunk 5 above)
Comment (Lumyo) [reply package.json:7]: Will be taken care with the plugin https://www.npmjs.com/package/hardhat-typechain

File: hardhat.config.ts (hunk 4 above)
Comment (Lumyo) [reply hardhat.config.ts:47]: Prettier doesnt have an option to add padding inside template litterals, only for brackets of objects

File: package.json (hunk 6 above)
Comment (Lumyo) [reply package.json:11]: I feel like eslint is redundant when using typescript, as typescript compiler already detects code errors (+ type errors which eslint doesnt check) and prettier takes care of the code formatting.

File: test/factory/CloneFactory.ts (hunk 8 above)
Comment (Lumyo) [reply test/factory/CloneFactory.ts:28]: Yes, prettier.

File: spec/token/ERC20/ERC20Base.behavior.ts (hunk 9 above)
Comment (Lumyo) [reply spec/token/ERC20/ERC20Base.behavior.ts:33]: Good point.
So by default, types with method signatures are only generated when there is an overload, but typechain has a `--always-generate-overloads` options, which generate types with method signature for everything.

So I'll update to use this flag, and will use everywhere the function call through the method signature, which will solve this.

File: contracts/signature/ERC1271Stored.sol (hunk 1 above)
Comment (ItsNickBarry) [reply contracts/signature/ERC1271Stored.sol:14]: Do you know of a fundamental reason for this, or should I open an issue in the TypeChain repository?

Comment (ItsNickBarry) [reply contracts/signature/ERC1271Stored.sol:14]: For now I think the solution is to add a line `signature = signature`.  I don't think this costs gas if the optimizer is turned on, though not sure.

File: test/factory/CloneFactory.ts (hunk 8 above)
Comment (ItsNickBarry) [reply test/factory/CloneFactory.ts:28]: Has prettier formatting been applied to all files?  Some are worse than others.

**Comment by ItsNickBarry** [comment]
@Lumyo How will this affect publication to NPM?

File: package.json (hunk 6 above)
Comment (ItsNickBarry) [reply package.json:11]: I think you'd be right, if prettier were actually capable of making things prettier.  For now the removal of eslint is fine, as it reduces complexity, but I might add it again at some point.

File: contracts/signature/ERC1271Stored.sol (hunk 1 above)
Comment (Lumyo) [reply contracts/signature/ERC1271Stored.sol:14]: No, not aware of a particular reason for this

File: test/factory/CloneFactory.ts (hunk 8 above)
Comment (Lumyo) [reply test/factory/CloneFactory.ts:28]: Yes it has

**Comment by ItsNickBarry** [comment]
Found a workaround to the failing Factory tests.  Underlying issue reported here: https://github.com/nomiclabs/hardhat/issues/1488

**Comment by Lumyo** [comment]
@ItsNickBarry 
I think this is the cleanest way to handle the TS build : https://github.com/solidstate-network/solidstate-solidity/pull/19/commits/e7d5978cefcc5e34f1c8c8b580734725f11498ac

- Moved `typechain` generation to `spec/typechain`
- Created a typescript path mapping from `@solidstate/typechain` to `spec/typechain`

Everywhere, we import typechain classes through `@solidstate/typechain` but this is not a separate package, so no need to publish it / link it for local development. The typescript compiler will remap the paths on compilation, and they will directly be included in `@solidstate/spec` which is the only published package which needs them.

If later we want to deploy a separate `@solidstate/typechain` package, it will be easy as the imports are done like that everywhere already (Would just need to remove the TS path mapping)

**Comment by Lumyo** [comment]
Actually, despite what i was thinking,, the path alias mentioned above are not remapped on compilation, so I'm gonna use relative import in `spec`

**Comment by ItsNickBarry** [comment]
Added a types package here: https://github.com/solidstate-network/solidstate-solidity/pull/22

The compilation override to copy the `typechain` directory isn't clean, but I had to do the same thing for the ABI package anyway.

**Comment by Lumyo** [comment]
@ItsNickBarry Can we merge this ?

**Comment by ItsNickBarry** [comment]
If all internal packages are set up to be published on NPM and imported by both JS and TS, then yes.  Can you confirm?  I was going to test by publishing to a local package registry like verdaccio.

**Comment by Lumyo** [comment]
`spec` and `library` packages typescript configs should be fine.
Just compile both using `tsc` and publish including the `dist` folder, which contain JS code + types (so will work with both JS and TS)
Not sure if there is anything else specific to NPM which needs to be done

**Pull request merged**